core/src/main/java/org/geysermc/geyser/translator/text/MessageTranslator.java

```diff
@@ -428,15 +433,95 @@ public static String normalizeSpace(String string) {
+    public static String deserializeDescription(GeyserSession session, NbtMap tag) {
         NbtMap description = tag.getCompound("description");
-        String translate = description.getString("translate", null);
-        if (translate == null) {
-            GeyserImpl.getInstance().getLogger().debug("Don't know how to read description! " + tag);
-            return "";
+        Component parsed = componentFromNbtTag(description);
+        return convertMessage(session, parsed);
+    }
+
+    public static Component componentFromNbtTag(Object nbtTag) {
+        return componentFromNbtTag(nbtTag, Style.empty());
+    }
+
+    private static Component componentFromNbtTag(Object nbtTag, Style style) {
+        if (nbtTag instanceof String literal) {
+            return Component.text(literal).style(style);
+        } else if (nbtTag instanceof List<?> list) {
+            return Component.join(JoinConfiguration.builder(), componentsFromNbtList(list, style));
+        } else if (nbtTag instanceof NbtMap map) {
+            Component component = Component.empty();
+            if (map.containsKey("text")) {
+                component = Component.text(map.getString("text"));
+            } else if (map.containsKey("translate")) {
+                String key = map.getString("translate");
+                String fallback = map.getString("fallback", "");
+                List<Component> args = new ArrayList<>();
+
+                if (map.containsKey("with")) {
+                    Object with = map.get("with");
+                    if (with instanceof List<?> list) {
+                        args = componentsFromNbtList(list, style);
+                    } else {
+                        args.add(componentFromNbtTag(with, style));
+                    }
+                }
+                component = Component.translatable(key, fallback, args);
+            }
+
+            Style newStyle = getStyleFromNbtMap(map, style);
+            component = component.style(newStyle);
+
+            Object extra = map.get("extra");
+            if (extra != null) {
+                component = component.append(componentFromNbtTag(extra, newStyle));
+            }
+
+            return component;
+        }
+
+        throw new IllegalArgumentException("Expected tag to be a literal string, a list of components, or a component");
+    }
+
+    private static List<Component> componentsFromNbtList(List<?> list, Style style) {
+        List<Component> components = new ArrayList<>();
+        for (Object entry : list) {
+            components.add(componentFromNbtTag(entry, style));
         }
-        return translate;
+        return components;
+    }
+
+    public static Style getStyleFromNbtMap(NbtMap map) {
+        return getStyleFromNbtMap(map, Style.empty());
+    }
+
+    public static Style getStyleFromNbtMap(NbtMap map, Style base) {
+        Style.Builder newStyle = Style.style().merge(base);
```

Comment:
I'm not sure if it's worth the processing tradeoff, but we could have `getStyleFromNbtMap0` which both this method and the above call - the `0` method can avoid merging an empty style from the above method. Does that make sense?

Reply:
Ended up making `getStyleFromNbtMap(NbtMap map)` read the style immediately, and making `getStyleFromNbtMap(NbtMap map, Style base)` use the other method to read the style and then merge that onto the base style. This should improve the performance since an empty style is no longer merged.